Develop

ivanearisty opened the pull request:
# Pull Request

## Description

This pull request establishes the complete, foundational implementation of the component-based Mail Client application. It delivers a professional-grade Python project that interacts with the Gmail API, built on a philosophy of strict interface-implementation separation and dependency injection.

The PR includes four core, self-contained components: two abstract protocols (`message`, `mail_client_api`) and their corresponding concrete Gmail implementations (`gmail_message_impl`, `gmail_client_impl`). It also includes a full development toolchain with linting, formatting, type checking, and a multi-layered testing strategy (unit, integration, and E2E).

## Type of Change

Please select the type of change this PR represents:

- [ ] 🐛 Bug fix (non-breaking change which fixes an issue)
- [x] ✨ New feature (non-breaking change which adds functionality)
- [ ] 💥 Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] 📚 Documentation update
- [ ] 🔧 Code refactoring (no functional changes)
- [ ] ⚡ Performance improvement
- [ ] 🧪 Test coverage improvement
- [ ] 🔨 Build/CI configuration changes

## Related Issues

N/A

## Changes Made

### Components Modified
- [x] `message` - Message protocol
- [x] `mail_client_api` - Mail client protocol  
- [x] `gmail_message_impl` - Gmail message implementation
- [x] `gmail_client_impl` - Gmail client implementation
- [x] Root configuration (pyproject.toml, CI/CD, etc.)
- [x] Documentation
- [x] Tests

### Detailed Changes
- **Protocols**: Defined the abstract `Message` and `Client` protocols to establish the core API contracts for the system.
- **Gmail Implementations**: Created `GmailMessage` for parsing raw Gmail API data and `GmailClient` for handling OAuth2 authentication and API interactions (fetching, deleting, marking as read).
- **Dependency Injection**: Established a factory pattern where importing implementation packages (`gmail_client_impl`, `gmail_message_impl`) automatically overrides the protocol's factory function.
- **Project Scaffolding**: Configured the top-level `pyproject.toml` as a `uv` workspace, defining all development dependencies, test markers, and quality tools (`ruff`, `mypy`).
- **Testing Infrastructure**: Implemented a comprehensive test suite with unit tests for each component, integration tests for component interaction, and end-to-end tests for the main application workflow. Added pytest markers (`unit`, `integration`, `e2e`, `circleci`, `local_credentials`) to categorize tests.
- **Documentation**: Created a complete documentation site using MkDocs and `mkdocstrings` to automatically generate API references from docstrings.

## Testing

### Test Categories Covered
- [x] Unit tests (`src/*/tests/`)
- [x] Integration tests (`tests/integration/`)
- [x] End-to-end tests (`tests/e2e/`)
- [x] CircleCI compatibility tests

### Testing Commands Used
```bash
# Run all unit tests for fast feedback
uv run pytest src/ -m "not local_credentials" -v

# Run integration and E2E tests which require credentials
uv run pytest tests/ -v

# Check for linting and formatting issues
uv run ruff check .
uv run ruff format --check .

# Run static type checking
uv run mypy src tests
```

### Test Results
- [x] All existing tests pass
- [x] New tests added for new functionality
- [x] Code coverage maintained/improved
- [x] No linting errors
- [x] Type checking passes

## Code Quality Checklist

### Component Architecture
- [x] Changes follow the interface/implementation separation principle
- [x] Dependency injection pattern maintained where applicable
- [x] Protocol contracts not broken
- [x] Components remain "forkliftable" (self-contained)

### Code Standards  
- [x] Code follows project style guidelines (ruff formatting)
- [x] All functions and classes have docstrings
- [x] Type hints are present and accurate
- [x] No unused imports or variables
- [x] Error handling is appropriate

### Documentation
- [x] README.md files updated if component APIs changed
- [x] Code comments added for complex logic
- [x] Component interaction documentation updated if needed

## CircleCI Status

- [x] CircleCI build passes
- [x] All quality gates met (linting, type checking, testing)
- [x] Code coverage threshold maintained

**CircleCI Build Link:** [Paste link to successful CircleCI build]

**Coverage Report Link:** [Paste link to coverage report artifact]

## Authentication Testing

### Credentials Used (if applicable)
- [x] Tested with local credential files (`credentials.json`, `token.json`)
- [x] Tested with environment variables (CI/CD mode)
- [x] Interactive authentication flow tested
- [x] Non-interactive authentication flow tested

### Authentication Scenarios
- [x] First-time OAuth2 setup works
- [x] Token refresh works correctly
- [x] Graceful handling of missing/invalid credentials
- [x] CI/CD environment compatibility verified

## Breaking Changes

This is the initial implementation of the project; there are no existing functionalities to break.

## Additional Context

### Performance Impact
- [x] No performance regression
- **Performance consideration:** The `gmail_client_impl` uses an iterator for `get_messages` to avoid loading all messages into memory at once. The `gmail_message_impl` uses lazy parsing for message properties.

### Dependencies
- [x] New dependencies justified and documented
- [x] Dependencies are pinned to specific versions in `uv.lock`

### Security Considerations
- [x] No sensitive data exposed
- [x] Authentication flows secure
- [x] API credentials properly handled (via `.gitignore` for local files and environment variables for CI/CD)

## Screenshots (if applicable)

N/A (This is a backend/library project with no UI changes).

## Reviewer Guidelines

### Focus Areas
Please pay special attention to:
- [x] Component boundaries and interface contracts
- [x] Authentication and credential handling
- [x] Test coverage and quality
- [x] Documentation accuracy

### Testing Instructions
1. Check out this branch
2. Run `uv sync --dev` to install dependencies
3. Run the testing commands listed above
4. Verify the changes work as described

## Definition of Done

- [x] Code follows project standards and passes all quality checks
- [x] Tests are comprehensive and pass
- [x] Documentation is updated and accurate
- [x] CircleCI build is successful
- [x] Changes are backward compatible OR migration path is documented
- [x] Security implications have been considered
- [x] Performance impact is acceptable
- [x] Component architecture principles are maintained

ivanearisty committed:
initialized repository members

ivanearisty committed:
Adding mkdocs auto documentation settings

ivanearisty committed:
Coverage setup + tests to mail_client_api

ivanearisty committed:
Gmail impl component setup + dep injection

ivanearisty committed:
Fix Automated Documentation for Gmail Client

ivanearisty committed:
Add interactive flow, add example secrets to README.md

ivanearisty committed:
Ruff fixes prior to development effort

ivanearisty committed:
MyPy fixes prior to development effort

ivanearisty committed:
Add init + dependency injection tests

ivanearisty committed:
Partial MyPy fixes

ivanearisty committed:
refactor message to be its own component

ivanearisty committed:
get messages method working

ivanearisty committed:
get other impl methods working

ivanearisty committed:
adding breaking tests

ivanearisty committed:
more test files

ivanearisty committed:
mypy + ruff fixes

ivanearisty committed:
add documentation

ivanearisty committed:
add circleci

ivanearisty committed:
small circleci fix

ivanearisty committed:
fix: Mock to_json() method in test_init_with_custom_token_uri

- Fixed TypeError in gmail client authentication test
- Mock credentials now properly returns JSON string from to_json()
- Resolves CircleCI test failure in unit test suite

ivanearisty committed:
fix: Disable coverage for integration tests in CircleCI

- Add --no-cov flag to integration_test job to prevent coverage failures
- Integration tests focus on component interactions, not code coverage
- Add explicit coverage settings to circleci_test job for clarity
- circleci_test job maintains coverage requirements since it includes unit tests

Resolves coverage failure when running integration tests in isolation

ivanearisty committed:
Test Consolidation

amoghkr1shna commented on .github/ISSUE_TEMPLATE/bug_report.md:
Nice addition

amoghkr1shna commented on .circleci/config.yml:
```diff
@@ -0,0 +1,292 @@
+          name: "Verify Installation (Optional)"
+          command: |
+            source .venv/bin/activate
+            uv pip list # Check if workspace members and tools are listed
```
I'm not sure about using `uv pip` from what was discussed in class

ivanearisty replied:
True, fixing

amoghkr1shna commented on README.md:
```diff
@@ -0,0 +1,222 @@
+3.  **`mail_client_api`**: Defines the abstract `Client` protocol. This is the contract for what actions a mail client can perform (e.g., `get_messages`).
+4.  **`gmail_client_impl`**: Provides the `GmailClient` class, a concrete implementation that uses the Google API to perform the actions defined in the `Client` protocol.
+
+## Project Structure
```
Structure is accurate

amoghkr1shna commented on docs/circleci-setup.md:
This file is empty - is the relevant CircleCI information added elsewhere?

ivanearisty replied:
forgot! thanks for the catch

amoghkr1shna commented:
- Documentation is meticulous, adding `feature_request` and `bug_report` are good additions. The API Documentation is also solid
- The Liskov Substitution mention was very interesting
- Having edge case testing is a solid move
- Even testing the CircleCI in the e2e testing is something I might lift from this PR
-  All in all, this is a solid email client implementation. The documentation is clear and well-structured. The folder structure is logical, and the separation of concerns between components is well-maintained.

ivanearisty committed:
Add project badges and modernize dependency configuration

- Add SVG badges to README.md for CircleCI, coverage, Python version, and code style
- Update pyproject.toml to use [project.optional-dependencies] instead of legacy [dependency-groups]
- This resolves uv sync --extra dev compatibility issues

ivanearisty committed:
Refine __init__.py docstrings to focus on public API and dependency injection

- Update gmail_client_impl/__init__.py docstring to focus on dependency injection role
- Update gmail_message_impl/__init__.py docstring to focus on module exports
- Docstrings now match the file scope rather than describing entire package implementation
- Emphasize the key role of overriding factory functions for dependency injection

ivanearisty committed:
revert docstring changes

ivanearisty committed:
Replace uv pip commands with pure uv commands in CircleCI config

- Replace 'uv pip list' with 'uv tree' for dependency verification
- Replace 'uv pip install types-requests' with 'uv add types-requests'
- Aligns with uv best practices and avoids mixing pip-style commands

ivanearisty committed:
Add comprehensive CircleCI setup documentation

- Document complete CI/CD pipeline setup and configuration
- Include environment variable setup with Gmail API credentials
- Explain workflow structure and job responsibilities
- Provide troubleshooting guide for common issues
- Add local development commands and security considerations
- Include maintenance procedures for dependencies and tests

ivanearisty committed:
summarizing circleci

ivanearisty committed:
Fix CircleCI build by adding --extra dev to uv sync

- Add --extra dev flag to uv sync command to install development dependencies
- This ensures ruff, pytest, mypy and other dev tools are available in CI
- Resolves 'ruff: command not found' error in lint job

ivanearisty committed:
Update documentation to use proper uv commands

- Replace direct pytest/ruff/mypy calls with 'uv run' prefix
- Fix dependency installation command to use --extra dev instead of --dev
- Ensure all documentation reflects consistent uv usage patterns
- Update both README.md and docs/testing.md for consistency

ivanearisty committed:
Very small linting error fixes

ivanearisty commented:
Merging and closing PR due to assignment submission deadline.

ivanearisty commented:
Closing PR due to assignment deadline. Please pm through slack if it needs to be re-opened
